subsys/net/lib/http/http_client.c

```diff
@@ -34,7 +34,19 @@ static int sendall(int sock, const void *buf, size_t len)
 	while (len) {
 		ssize_t out_len = zsock_send(sock, buf, len, 0);
 
-		if (out_len < 0) {
+		if ((out_len == 0) || (out_len < 0 && errno == EAGAIN)) {
+			struct zsock_pollfd pfd;
+			int pollres;
+
+			pfd.fd = sock;
+			pfd.events = ZSOCK_POLLOUT;
+			pollres = zsock_poll(&pfd, 1, CONFIG_HTTP_CLIENT_SOCKET_POLL_TIMEOUT_MS);
```

Comment:
The `http_client_req()` API call has timeout option, and I think we should take that value into account here. So if the caller has set a timeout then we could do `req->internal.timeout` to wait max amount of time. If user has not set timeout (so that we do not wait), then we should not do any wait here either. Also the timeout should count both send and recv parts, so if we wait here some time, then the wait time in recv should be smaller.

Reply:
Currently, timeout is enforced via closing the socket by the timer anyway.
Which will cause `zsock_poll `to return, regardless of its timeout. 

But yes, I agree that it doesn't make any sense to have bigger timeout in zsock_poll, so we could add logic to use smaller of timeouts:

```
int timeout = (req_timeout < CONFIG_HTTP_CLIENT_SOCKET_POLL_TIMEOUT_MS) ? req_timeout : CONFIG_HTTP_CLIENT_SOCKET_POLL_TIMEOUT_MS;
pollres = zsock_poll(&pfd, 1, req_timeout);
```

What do you think?

Reply:
@jukkar Could you please have a look at the point above?
Please let me know if that's what you had it mind

Reply:
Sorry for late reply, I just noticed your question.

What I meant that we should initialize the req_timeout to the value that the user is supplying, and then just decrement it until it is 0. This means that we should never exceed the timeout value that the user has supplied in `http_client_req()` call.